nasa/ow_autonomy
# PLEXIL interface to joint telemetry

### Summary

 - Implemented unified interface for joint telemetries: position, effort/torque, velocity, acceleration.

### Supporting Changes

- Revamped existing representation and management of joint telemetry in PLEXIL adapter.

### Future work

- Much repetitive coding that could be streamlined.
- Antenna pan/tilt lookups could be implemented in the new interface rather than be separate.

### Test

Two tests are needed, for OceanWATERS and OWLAT.  First do a clean build that also builds for OWLAT (`catkin build -DOWLAT=ON`).  Re-source the OS init file each shell.

1. OceanWATERS
 - Start the Atacama world
 - Run the `TestTelemetry` plan.  Expected output is below.
 - Terminate both nodes for the OWLAT run.
```
[ INFO]: PLEXIL: Pan degees: -0.000374146164403517
[ INFO]: PLEXIL: Pan radians: -6.53011105811174e-06
[ INFO]: PLEXIL: Tilt degees: -0.671719016855934
[ INFO]: PLEXIL: Tilt radians: -0.0117237084923953
[ INFO]: PLEXIL: Position of joint 0 = -6.53011105811174e-06
[ INFO]: PLEXIL: Velocity of joint 0 = -2.41949916810734e-09
[ INFO]: PLEXIL: Effort of joint 0 = -5.16886089485524e-08
[ WARN]: jointTelemetry: acceleration not available for antenna joint.
[ INFO]: PLEXIL: Acceleration of joint 0 = 0
[ INFO]: PLEXIL: Position of joint 1 = -0.0117237084923953
[ INFO]: PLEXIL: Velocity of joint 1 = -3.05800119775946e-09
[ INFO]: PLEXIL: Effort of joint 1 = 2.37190864632781
[ WARN]: jointTelemetry: acceleration not available for antenna joint.
[ INFO]: PLEXIL: Acceleration of joint 1 = 0
[ INFO]: PLEXIL: Position of joint 2 = 2.88909033861065
[ INFO]: PLEXIL: Velocity of joint 2 = -6.43104181733693e-19
[ INFO]: PLEXIL: Effort of joint 2 = 1.82882900637571
[ INFO]: PLEXIL: Acceleration of joint 2 = -4.90600624849805e-16
[ INFO]: PLEXIL: Position of joint 3 = -0.120415069096364
[ INFO]: PLEXIL: Velocity of joint 3 = -0.00185637261719489
[ INFO]: PLEXIL: Effort of joint 3 = 0
[ INFO]: PLEXIL: Acceleration of joint 3 = 7.18058973431094e-05
[ INFO]: PLEXIL: Position of joint 4 = -1.24344978758018e-14
[ INFO]: PLEXIL: Velocity of joint 4 = 2.14554665630918e-18
[ INFO]: PLEXIL: Effort of joint 4 = 1.24355270296518e-11
[ INFO]: PLEXIL: Acceleration of joint 4 = 1.41459351537809e-16
[ INFO]: PLEXIL: Position of joint 5 = 3.60139097927089
[ INFO]: PLEXIL: Velocity of joint 5 = 2.01341640281611e-17
[ INFO]: PLEXIL: Effort of joint 5 = 5.15696364889991
[ INFO]: PLEXIL: Acceleration of joint 5 = 3.05818482822919e-15
[ INFO]: PLEXIL: Position of joint 6 = 8.88178419700125e-16
[ INFO]: PLEXIL: Velocity of joint 6 = 1.7969177386686e-18
[ INFO]: PLEXIL: Effort of joint 6 = 1.37298359110987e-17
[ INFO]: PLEXIL: Acceleration of joint 6 = 8.25443378170301e-17
[ INFO]: PLEXIL: Position of joint 7 = 1.55654376610611
[ INFO]: PLEXIL: Velocity of joint 7 = -1.41119465092063e-16
[ INFO]: PLEXIL: Effort of joint 7 = 3.10665754099172
[ INFO]: PLEXIL: Acceleration of joint 7 = -1.30032810491309e-14
[ INFO]: PLEXIL: Position of joint 8 = -1.49999999999678
[ INFO]: PLEXIL: Velocity of joint 8 = -2.50808633234961e-17
[ INFO]: PLEXIL: Effort of joint 8 = 6.80961724848328e-12
[ INFO]: PLEXIL: Acceleration of joint 8 = 2.02732222915155e-15
```

2. OWLAT test
 - Start roscore
 - `launch_owlat Scoop`
 - Launch owlat_exec and run the TestTelemetry plan.  E.g. `roslaunch ow_plexil owlat_exec.launch plan:=TelemetryTest.plx`.  Output should be as follows.

```
[ INFO]: Starting PLEXIL executive node...
[ INFO]: Executive node started, ready for PLEXIL plans.
[ INFO]: PLEXIL: Pan degees: 0
[ INFO]: PLEXIL: Pan radians: 0
[ INFO]: PLEXIL: Tilt degees: 0
[ INFO]: PLEXIL: Tilt radians: 0
[ INFO]: PLEXIL: Position of joint 0 = -6.68719914197442e-07
[ INFO]: PLEXIL: Velocity of joint 0 = 0
[ INFO]: PLEXIL: Effort of joint 0 = 0
[ INFO]: PLEXIL: Acceleration of joint 0 = 0
[ INFO]: PLEXIL: Position of joint 1 = -1.99000421035986
[ INFO]: PLEXIL: Velocity of joint 1 = 1.12020177876595e-06
[ INFO]: PLEXIL: Effort of joint 1 = 0
[ INFO]: PLEXIL: Acceleration of joint 1 = 0
[ INFO]: PLEXIL: Position of joint 2 = -3.86771435772015e-08
[ INFO]: PLEXIL: Velocity of joint 2 = 9.0606469559997e-09
[ INFO]: PLEXIL: Effort of joint 2 = 0.0195286236618256
[ INFO]: PLEXIL: Acceleration of joint 2 = -2.10235034485418e-09
[ INFO]: PLEXIL: Position of joint 3 = 3.10003013951662
[ INFO]: PLEXIL: Velocity of joint 3 = -3.75286495544945e-08
[ INFO]: PLEXIL: Effort of joint 3 = -0.115904641858132
[ INFO]: PLEXIL: Acceleration of joint 3 = 2.35063168929628e-10
[ INFO]: PLEXIL: Position of joint 4 = 1.38054097649685e-08
[ INFO]: PLEXIL: Velocity of joint 4 = -3.59024678005457e-09
[ INFO]: PLEXIL: Effort of joint 4 = -0.0191918291123874
[ INFO]: PLEXIL: Acceleration of joint 4 = 6.11062489497272e-10
[ INFO]: PLEXIL: Position of joint 5 = 1.84802149969342e-07
[ INFO]: PLEXIL: Velocity of joint 5 = -4.79501136214769e-08
[ INFO]: PLEXIL: Effort of joint 5 = -0.25336554980601
[ INFO]: PLEXIL: Acceleration of joint 5 = 1.17776565389249e-08
[ERROR]: getJointTelemetry: invalid joint index 6
[ INFO]: PLEXIL: Position of joint 6 = 0
[ERROR]: getJointTelemetry: invalid joint index 6
[ INFO]: PLEXIL: Velocity of joint 6 = 0
[ERROR]: getJointTelemetry: invalid joint index 6
[ INFO]: PLEXIL: Effort of joint 6 = 0
[ERROR]: getJointTelemetry: invalid joint index 6
[ INFO]: PLEXIL: Acceleration of joint 6 = 0
[ERROR]: getJointTelemetry: invalid joint index 7
[ INFO]: PLEXIL: Position of joint 7 = 0
[ERROR]: getJointTelemetry: invalid joint index 7
[ INFO]: PLEXIL: Velocity of joint 7 = 0
[ERROR]: getJointTelemetry: invalid joint index 7
[ INFO]: PLEXIL: Effort of joint 7 = 0
[ERROR]: getJointTelemetry: invalid joint index 7
[ INFO]: PLEXIL: Acceleration of joint 7 = 0
[ERROR]: getJointTelemetry: invalid joint index 8
[ INFO]: PLEXIL: Position of joint 8 = 0
[ERROR]: getJointTelemetry: invalid joint index 8
[ INFO]: PLEXIL: Velocity of joint 8 = 0
[ERROR]: getJointTelemetry: invalid joint index 8
[ INFO]: PLEXIL: Effort of joint 8 = 0
[ERROR]: getJointTelemetry: invalid joint index 8
[ INFO]: PLEXIL: Acceleration of joint 8 = 0
```
